src/smc-webapp/assistant/body.cjsx

```diff
@@ -0,0 +1,202 @@
+{React, ReactDOM, redux, Redux, Actions, Store, rtypes, rclass} = require('../smc-react')
+{Loading, Icon, Markdown, Space} = require('../r_misc')
+# cocalc libs
+{defaults, required, optional} = misc = require('smc-util/misc')
+
+
+exports.ExamplesBody = rclass
+    displayName : 'ExamplesBody'
+
+    propTypes:
+        actions             : rtypes.object
+        data                : rtypes.immutable
+        lang                : rtypes.string
+        code                : rtypes.string
+        descr               : rtypes.string
+        setup_code          : rtypes.string
+        prepend_setup_code  : rtypes.bool
+        cat0                : rtypes.number
+        cat1                : rtypes.number
+        cat2                : rtypes.number
+        catlist0            : rtypes.arrayOf(rtypes.string)
+        catlist1            : rtypes.arrayOf(rtypes.string)
+        catlist2            : rtypes.arrayOf(rtypes.string)
+        search_str          : rtypes.string
+        search_sel          : rtypes.number
+        hits                : rtypes.arrayOf(rtypes.array)
+        unknown_lang        : rtypes.bool   # if true, show info about contributing to the assistant
+
+    getDefaultProps: ->
+        search_str : ''
+
+    shouldComponentUpdate: (props, state) ->
+        ret = misc.is_different(@props, props, [
+            'data', 'lang', 'code', 'descr', 'setup_code', 'prepend_setup_code',
+            'cat0', 'cat1', 'cat2', 'search_str', 'search_sel', 'unknown_lang'
+        ])
+        ret or= misc.is_different_array(props.hits, @props.hits)
+        ret or= misc.is_different_array(props.catlist0, @props.catlist0)
+        ret or= misc.is_different_array(props.catlist1, @props.catlist1)
+        ret or= misc.is_different_array(props.catlist2, @props.catlist2)
+        return ret
+
+    componentDidMount: ->
+        @scrollTo0 = _.debounce((() -> $(ReactDOM.findDOMNode(@refs.list_0)).find('.active').scrollintoview()), 50)
+        @scrollTo1 = _.debounce((() -> $(ReactDOM.findDOMNode(@refs.list_1)).find('.active').scrollintoview()), 50)
+        @scrollTo2 = _.debounce((() -> $(ReactDOM.findDOMNode(@refs.list_2)).find('.active').scrollintoview()), 50)
+        @scrollToS = _.debounce((() -> $(ReactDOM.findDOMNode(@refs.search_results_list)).find('.active').scrollintoview()), 50)
+
+    componentDidUpdate: (props, state) ->
+        @scrollTo0() if props.cat0 != @props.cat0
+        @scrollTo1() if props.cat1 != @props.cat1
+        @scrollTo2() if props.cat2 != @props.cat2
+        @scrollToS() if props.search_sel != @props.search_sel
+
+    category_selection: (level, idx) ->
+        @props.actions.set_selected_category(level, idx)
+
+    # level could be 0, 1 or 2
+    category_list: (level) ->
+        cat  = @props["cat#{level}"]
+        list = @props["catlist#{level}"]
```

Comment:
2 comments earlier I started to make this a list of lists to avoid this, too. Ended up to be too confusing, so not a good idea... changing this now this way, yes.